File: svc-data/src/main/java/com/beancounter/marketdata/trn/TrnService.java
package com.beancounter.marketdata.trn;

import com.beancounter.common.contracts.PortfoliosResponse;
import com.beancounter.common.contracts.TrnRequest;
import com.beancounter.common.contracts.TrnResponse;
import com.beancounter.common.exception.BusinessException;
import com.beancounter.common.model.Portfolio;
import com.beancounter.common.model.Trn;
import com.beancounter.common.utils.DateUtils;
import com.beancounter.marketdata.portfolio.PortfolioService;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Optional;
import javax.transaction.Transactional;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

@Slf4j
@Service
@Transactional
public class TrnService {
  private final TrnRepository trnRepository;
  private final TrnAdapter trnAdapter;
  private final PortfolioService portfolioService;
  private final DateUtils dateUtils = new DateUtils();

  TrnService(TrnRepository trnRepository,
             TrnAdapter trnAdapter,
             PortfolioService portfolioService) {
    this.trnRepository = trnRepository;
    this.portfolioService = portfolioService;
    this.trnAdapter = trnAdapter;
  }

  public PortfoliosResponse findWhereHeld(String assetId, LocalDate tradeDate) {
    if (tradeDate == null) {
      tradeDate = dateUtils.getDate(dateUtils.today());
    }
    Collection<Portfolio> portfolios = trnRepository
        .findDistinctPortfolioByAssetIdAndTradeDate(assetId, tradeDate);
    log.info("Found {}", portfolios.size());
    return PortfoliosResponse.builder().data(portfolios).build();
  }

  public TrnResponse getPortfolioTrn(Portfolio portfolio, String trnId) {
    Optional<Trn> trn = trnRepository.findByPortfolioIdAndId(portfolio.getId(), trnId);
    Optional<TrnResponse> result = trn.map(
        transaction -> hydrate(Collections.singleton(transaction)));

    if (result.isEmpty()) {
      throw new BusinessException(String.format("Trn %s not found", trnId));
    }
    return result.get();

  }

  public TrnResponse save(Portfolio portfolio, TrnRequest trnRequest) {
    TrnResponse trnResponse = trnAdapter.convert(portfolio, trnRequest);
    Iterable<Trn> saved = trnRepository.saveAll(trnResponse.getData());
    Collection<Trn> trns = new ArrayList<>();
    saved.forEach(trns::add);
    trnResponse.setData(trns);
    log.trace("Wrote {}/{} transactions for {}",
        trnResponse.getData().size(), trnRequest.getData().size(), portfolio.getCode());

    return trnResponse;
  }

  /**
   * Display order.
   *
   * @param portfolio fully qualified portfolio the caller is authorised to view
   * @param assetId filter by pk
   * @return Transactions in display order that is friendly for viewing.
   */
  public TrnResponse findByPortfolioAsset(Portfolio portfolio, String assetId) {
    Collection<Trn> results = trnRepository
        .findByPortfolioIdAndAssetId(
            portfolio.getId(),
            assetId, Sort.by("asset.code")
                .and(Sort.by("tradeDate").descending()));

    log.debug("Found {} for portfolio {} and asset {}",
        results.size(),
        portfolio.getCode(),
        assetId
    );
    return hydrate(results, true);

  }

  /**
   * Processing order.
   *
   * @param portfolio trusted
   * @param assetId   filter by pk
   * @param tradeDate until this date inclusive
   * @return transactions that can be accumulated into a position
   */
  public TrnResponse findByPortfolioAsset(
      Portfolio portfolio,
      String assetId,
      LocalDate tradeDate) {

    Collection<Trn> results = trnRepository
        .findByPortfolioIdAndAssetIdUpTo(
            portfolio.getId(),
            assetId,
            tradeDate);

    log.debug("Found {} for portfolio {} and asset {}",
        results.size(),
        portfolio.getCode(),
        assetId
    );
    return hydrate(results, false);
  }

  public TrnResponse findForPortfolio(Portfolio portfolio) {
    Collection<Trn> results = trnRepository.findByPortfolioId(portfolio.getId(),
        Sort.by("asset.code")
            .and(Sort.by("tradeDate")));
    log.debug("trns: {}, portfolio: {}", results.size(), portfolio.getCode());
    return hydrate(results);
  }

  /**
   * Purge transactions for a portfolio.
   *
   * @param portfolio portfolio owned by the caller
   * @return number of deleted transactions
   */
  public long purge(Portfolio portfolio) {
    return trnRepository.deleteByPortfolioId(portfolio.getId());
  }

  private Trn setAssets(Trn trn) {
    trn.setAsset(trnAdapter.hydrate(trn.getAsset()));
    trn.setCashAsset(trnAdapter.hydrate(trn.getCashAsset()));
    return trn;
  }

  private TrnResponse hydrate(Iterable<Trn> trns) {
    return hydrate(trns, true);
  }

  private TrnResponse hydrate(Iterable<Trn> trns, boolean secure) {
    Collection<Trn> results = new ArrayList<>();
    for (Trn trn : trns) {
      boolean add = (!secure || portfolioService.canView(trn.getPortfolio()));
      if (add) {
        results.add(setAssets(trn));
      }
    }
    if (results.isEmpty()) {
      return TrnResponse.builder().build(); // Empty
    }
    return TrnResponse.builder().data(results).build();
  }

}

File: svc-data/src/main/kotlin/com/beancounter/marketdata/trn/TrnService.kt
package com.beancounter.marketdata.trn

import com.beancounter.common.contracts.TrnDeleteResponse
import com.beancounter.common.contracts.TrnRequest
import com.beancounter.common.contracts.TrnResponse
import com.beancounter.common.contracts.TrnSaveResult
import com.beancounter.common.exception.NotFoundException
import com.beancounter.common.input.TrnInput
import com.beancounter.common.model.Portfolio
import com.beancounter.common.model.Trn
import com.beancounter.marketdata.cache.CacheInvalidationProducer
import com.beancounter.marketdata.cash.CashAutoSettleService
import com.beancounter.marketdata.portfolio.PortfolioAccessControl
import com.beancounter.marketdata.portfolio.PortfolioService
import jakarta.transaction.Transactional
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.time.LocalDate
import java.util.function.Consumer

/**
 * Core service for transaction writes (create / patch / delete / purge).
 *
 * Reads live in [TrnFinder] (portfolio-scoped) and [TrnQueryService] (asset /
 * event). Settle / unsettle lifecycle lives in [TrnSettlementService]. Shared
 * read-time processing lives in [TrnPostProcessor]. Broker operations live in
 * [TrnBrokerService]; investment/income analysis in [TrnAnalysisService].
 */
@Service
@Transactional
class TrnService(
    private val trnRepository: TrnRepository,
    private val trnInputMapper: TrnInputMapper,
    private val portfolioService: PortfolioService,
    private val portfolioAccessControl: PortfolioAccessControl,
    private val cacheInvalidationProducer: CacheInvalidationProducer,
    private val cashAutoSettleService: CashAutoSettleService,
    private val trnFinder: TrnFinder
) {
    private val log = LoggerFactory.getLogger(TrnService::class.java)

    fun save(
        portfolioId: String,
        trnRequest: TrnRequest
    ): Collection<Trn> {
        val portfolio = portfolioService.find(portfolioId)
        return save(portfolio, trnRequest)
    }

    fun save(
        portfolio: Portfolio,
        trnRequest: TrnRequest
    ): Collection<Trn> = saveWithResult(portfolio, trnRequest).trns

    fun saveWithResult(
        portfolioId: String,
        trnRequest: TrnRequest
    ): TrnSaveResult = saveWithResult(portfolioService.find(portfolioId), trnRequest)

    /**
     * Persist trns and return them along with any non-fatal warnings raised
     * during auto-settle (e.g. cash funding portfolio has no balance in the
     * trade currency). Controllers wrap the warnings into [TrnResponse] so
     * the UI can surface them; legacy callers (CashTransferService etc.)
     * use the [save] overload that only returns the trns.
     */
    fun saveWithResult(
        portfolio: Portfolio,
        trnRequest: TrnRequest
    ): TrnSaveResult {
        val saved =
            trnRepository.saveAll(
                trnInputMapper.convert(portfolio, trnRequest)
            )
        // Asset hydration happens via AssetEntityListener @PostLoad on every load —
        // saveAll returns entities whose asset reference was already hydrated upstream.
        val results: MutableCollection<Trn> = mutableListOf()
        saved.forEach(Consumer { e: Trn -> results.add(e) })
        // Auto-settle cash to the linked funding portfolio. Skips non-trigger
        // types (DEPOSIT/WITHDRAWAL etc.) — no recursion risk.
        val warnings = mutableListOf<String>()
        for (trn in results.toList()) {
            // Cash auto-settle emits a compensating pair whose status mirrors the
            // parent — a PROPOSED trade carries PROPOSED legs (no settled cash
            // impact), a SETTLED trade carries SETTLED legs. Settling later
            // reconciles them to SETTLED. emitCompensatingTransfer no-ops for
            // non-trigger types (DEPOSIT/WITHDRAWAL etc.), so no recursion.
            val res = cashAutoSettleService.emitCompensatingTransfer(trn)
            warnings.addAll(res.warnings)
        }
        if (trnRequest.data.size == 1) {
            log.trace(
                "Wrote 1 transaction asset: ${trnRequest.data[0].assetId}, portfolio: ${portfolio.code}"
            )
        } else {
            log.trace(
                "Wrote ${results.size}/${trnRequest.data.size} transactions for ${portfolio.code}"
            )
        }
        val earliestDate = results.minOfOrNull { it.tradeDate }
        if (earliestDate != null) {
            cacheInvalidationProducer.sendTransactionEvent(portfolio.id, earliestDate)
        }
        return TrnSaveResult(results.toList(), warnings.toList())
    }

    fun purge(portfolioId: String): Long {
        val portfolio = portfolioService.find(portfolioId)
        return purge(portfolio)
    }

    /**
     * Purge transactions for a portfolio.
     */
    fun purge(portfolio: Portfolio): Long {
        log.debug("Purging transactions for {}", portfolio.code)
        val count = trnRepository.deleteByPortfolioId(portfolio.id)
        cacheInvalidationProducer.sendTransactionEvent(portfolio.id, LocalDate.MIN)
        return count
    }

    fun delete(trnId: String): Collection<String> {
        val result = trnRepository.getOrThrow(trnId)
        val deleted = mutableListOf<Trn>()
        if (portfolioAccessControl.canView(result.portfolio)) {
            trnRepository.delete(result)
            deleted.add(result)
            cacheInvalidationProducer.sendTransactionEvent(result.portfolio.id, result.tradeDate)
        }
        return deleted.map { it.id }
    }

    /**
     * Delete one trn and surface its auto-settled siblings (without cascading).
     * The UI uses [TrnDeleteResponse.siblings] to prompt the user before
     * issuing follow-up DELETEs for the W+D cash legs.
     */
    fun deleteWithSiblings(trnId: String): TrnDeleteResponse {
        val parent = trnRepository.getOrThrow(trnId)
        if (!portfolioAccessControl.canView(parent.portfolio)) {
            // Match unsettle / delete-existing behaviour — surface as 404
            // rather than a silent empty success.
            throw NotFoundException("Transaction not found: $trnId")
        }
        val siblings = cashAutoSettleService.findSiblings(parent).map { it.id }
        trnRepository.delete(parent)
        cacheInvalidationProducer.sendTransactionEvent(parent.portfolio.id, parent.tradeDate)
        return TrnDeleteResponse(listOf(parent.id), siblings)
    }

    fun patch(
        portfolioId: String,
        trnId: String,
        trnInput: TrnInput
    ): TrnResponse {
        val portfolio = portfolioService.find(portfolioId)
        return patch(portfolio, trnId, trnInput)
    }

    fun patch(
        portfolio: Portfolio,
        trnId: String,
        trnInput: TrnInput
    ): TrnResponse {
        val existing = trnFinder.getPortfolioTrn(trnId)
        val trn =
            trnInputMapper.map(
                portfolio,
                trnInput,
                existing.iterator().next()
            )
        trnRepository.save(trn)
        // Re-sync the compensating cash transfer to the edited values — reconcile
        // deletes any stale pair and re-posts so the legs track the new
        // amount/date/currency AND the new status. Editing a SETTLED trade down to
        // PROPOSED must revert its legs to PROPOSED too (not orphan them settled).
        val warnings = cashAutoSettleService.emitCompensatingTransfer(trn).warnings
        cacheInvalidationProducer.sendTransactionEvent(portfolio.id, trn.tradeDate)
        return TrnResponse(arrayListOf(trn), warnings)
    }
}